src/hmm_functions.R

```diff
@@ -181,18 +182,24 @@ get_hmm_and_minimum_distance_estimates_random_initialization <- function(params0
     dtable[, y_one_period_ahead := c(tail(y, .N-1), NA), by="point_id"]
     dtable[, y_two_periods_ahead := c(tail(y, .N-2), NA, NA), by="point_id"]
 
+    ## TODO How does this behave when Y is MCAR?
+    ## Joint distribution of (Y_{t+1}, Y_{t})
     M_Y_joint_hat_list <- lapply(seq_len(max(dtable$time) - 1), function(fixed_t) {
         with(subset(dtable, time == fixed_t), prop.table(table(y_one_period_ahead, y)))
-    })  # Joint distribution of (Y_{t+1}, Y_{t}) and (Y_{t+2}, Y_{t+1})
+    })
 
     ## Compute inverses once and pass them to get_min_distance_estimates / solnp
     M_Y_joint_hat_inverse_list <- lapply(M_Y_joint_hat_list, solve)
 
     ## TODO Need to handle edge case where any of these matrices are not invertible, which might happen at small sample sizes
+    ## Joint distribution of (Y_{t+2}, Y_{t+1}, Y_{t})
     M_fixed_y_Y_joint_hat_list <- lapply(seq_len(params0$n_components), function(fixed_y) {
         lapply(seq_len(max(dtable$time) - 2), function(fixed_t) {
             return(with(subset(dtable, time == fixed_t & y_two_periods_ahead == fixed_y),
-                        table(y_one_period_ahead, y)) / sum(dtable$time == fixed_t))
+                        table(y_one_period_ahead, y)) / sum(dtable$time == fixed_t &
+                                                            !is.na(dtable$y_two_periods_ahead) &
+                                                            !is.na(dtable$y_one_period_ahead) &
+                                                            !is.na(dtable$y)))
```

Comment:
What would this look like if we instead allowed Pr[Y | S] to be a non-square matrix, for example with Y in {0, 1, NA} and S in {0, 1}?

Reply:
That would allow us to estimate Pr[ Y is missing | S ] that varies with S, which may not be relevant in most applications.